ArduCopter/mode.cpp

```diff
@@ -397,38 +397,41 @@ void Copter::notify_flight_mode() {
 
 // get_pilot_desired_angle - transform pilot's roll or pitch input into a desired lean angle
 // returns desired angle in centi-degrees
-void Mode::get_pilot_desired_lean_angles(float &roll_out, float &pitch_out, float angle_max, float angle_limit) const
+void Mode::get_pilot_desired_lean_angles(float &roll_out_cd, float &pitch_out_cd, float angle_max_cd, float angle_limit_cd) const
 {
     // throttle failsafe check
     if (copter.failsafe.radio || !copter.ap.rc_receiver_present) {
-        roll_out = 0;
-        pitch_out = 0;
+        roll_out_cd = 0.0;
+        pitch_out_cd = 0.0;
         return;
     }
-    // fetch roll and pitch inputs
-    roll_out = channel_roll->get_control_in();
-    pitch_out = channel_pitch->get_control_in();
+    // fetch roll and pitch stick positions
+    float thrust_angle_x_cd = - channel_pitch->get_control_in();
+    float thrust_angle_y_cd = channel_roll->get_control_in();
 
     // limit max lean angle
-    angle_limit = constrain_float(angle_limit, 1000.0f, angle_max);
+    angle_limit_cd = constrain_float(angle_limit_cd, 1000.0f, angle_max_cd);
 
-    // scale roll and pitch inputs to ANGLE_MAX parameter range
-    float scaler = angle_max/(float)ROLL_PITCH_YAW_INPUT_MAX;
-    roll_out *= scaler;
-    pitch_out *= scaler;
+    // scale roll and pitch inputs to +- angle_max
+    float scaler = angle_max_cd/(float)ROLL_PITCH_YAW_INPUT_MAX;
+    thrust_angle_x_cd *= scaler;
+    thrust_angle_y_cd *= scaler;
 
-    // do circular limit
-    float total_in = norm(pitch_out, roll_out);
-    if (total_in > angle_limit) {
-        float ratio = angle_limit / total_in;
-        roll_out *= ratio;
-        pitch_out *= ratio;
+    // convert square mapping to circular mapping with maximum magnitude of angle_limit
+    float total_in = norm(thrust_angle_x_cd, thrust_angle_y_cd);
+    if (total_in > angle_limit_cd) {
+        float ratio = angle_limit_cd / total_in;
+        thrust_angle_x_cd *= ratio;
+        thrust_angle_y_cd *= ratio;
     }
 
-    // do lateral tilt to euler roll conversion
-    roll_out = (18000/M_PI) * atanf(cosf(pitch_out*(M_PI/18000))*tanf(roll_out*(M_PI/18000)));
+    // thrust_angle_x and thrust_angle_y represents a level body frame thrust vector in the
+    // direction of [thrust_angle_x, thrust_angle_y] and a magnitude
+    // tan(mag([thrust_angle_x, thrust_angle_y])) * 9.81 * aircraft mass.
 
-    // roll_out and pitch_out are returned
+    // Conversion from angular thrust vector to euler angles.
+    roll_out_cd = (18000/M_PI) * atanf(cosf(thrust_angle_x_cd*(M_PI/18000))*tanf(thrust_angle_y_cd*(M_PI/18000)));
```

Comment:
so just to be clear, the thrust_angle_x_cd will be the opposite sign from what it was before this change.  Maybe the sign doesn't matter but just want to be sure.

Reply:
To get the orientation of thrust_angle_x_cd correct I add a negative sign on line 409.

This is used on lines 417, 421, 424, 433 and 434.

417 is only scaling thrust_angle_x_cd 
421 squares thrust_angle_x_cd  and therefore the sign does not matter
424 again only scales thrust_angle_x_cd 
433 takes the cos(thrust_angle_x_cd ). Given cos(-angle) = cos(angle) the sign does not matter.
finally on line 434 we set pitch_out_cd = - thrust_angle_x_cd; where we multiply again by -1 cancelling our initial multiplication by -1.